README.md
# Docker-Maniaplanet

This work is based on PyPlanet docker image (https://github.com/PyPlanet/maniaplanet-docker).

My goal is to be up-to-date and to be much more customisable.

- Use latest alpine image
- Use latest Maniaplanet server
- Lighter image
- Multi architectures (see below)
- Cleaner and more readable Dockerfile and entrypoint
- Every `dedicated_cfg` option is customisable through environment variable
- Pterodactyl compliant (https://github.com/pterodactyl/panel)

Works great with https://git.tools01.noxinmortus.fr/sysadmins/docker/docker-maniacontrol

This image is rootless.

## Official NoxInmortus repositories

Find more at :
- https://hub.docker.com/u/noxinmortus
- https://git.tools01.noxinmortus.fr/public
- https://github.com/NoxInmortus?tab=repositories

## Available Architectures
- amd64
- arm64 (aarch64)
- armv7 (arm)

## Common Usage
```
docker run --name=maniaplanet -d -v maniaplanet:/home/container --restart=unless-stopped \
  -p 2350:2350 -p 2350:2350/udp -p 3450:3450 -p 3450:3450/udp \
  -e MASTERSERVER_ACCOUNT='MyAccountName' \
  -e MASTERSERVER_ACCOUNT_PWD='P@55w0rd' \
  -e MASTERSERVER_KEY='yOuRkEy' \
  -e SUPERADMIN_PWD='AnotherP@55w0rd' \
  -e ADMIN_PWD='AnotherP@55w0rd' \
  -e SERVER_NAME='Aw3s0meS3rV3r' \
  -e MAX_PLAYERS='16' \
  -e MAX_SPECTATORS='6' \
  -e SERVER_PASSWORD='s0m371ng' \
 noxinmortus/docker-maniaplanet
```

Once started, you can also manually edit the `config.xml` as it will not be overwritten if the file exist. Same thing for the `matchsettings.xml`

## User / Group Identifiers

When using volumes (`-v` flags) permissions issues can arise between the host OS and the container, we avoid this issue by allowing you to specify the user `UID` and group `GID` :
```
-e UID=1000 # for UserID
-e GID=1000 # for GroupID
```

Ensure any volume directories on the host are owned by the same user you specify and any permissions issues will vanish like magic.

In this instance `UID=1000` and `GID=1000`, to find yours use `id user` as below:
```
  $ id username
    uid=1000(dockeruser) gid=1000(dockergroup) groups=1000(dockergroup)
```

## Configuration

Volume to mount is `/home/container` (to be Pterodactyl compliant) and contains all maniaplanet files.

|Ports|Usage|
|-|-|
|2350|Server TCP port|
|2350/udp|Server UDP port|
|3450|Server P2P TCP port|
|3450/udp|Server P2P UDP port|
|5000|XMLRPC Port|

|Environment variables|Default|Usage|
|-|-|-|
|MASTERSERVER_ACCOUNT|None||
|MASTERSERVER_ACCOUNT_PWD|None||
|MASTERSERVER_KEY|None|Validation key||
|FORCE_IP_ADDRESS|`$(wget -4 -qO- http://ifconfig.co)`||
|SERVER_PORT|`2350`||
|SERVER_P2P_PORT|`3450`||
|XMLRPC_PORT|`5000`||
|XMLRPC_ALLOW_REMOTE|`True`|If you specify an ip adress here, it'll be the only accepted adress. this will improve security|
|TITLE|`TMStadium@nadeo`|SMStorm, TMCanyon...|
|TITLE_PACK_URL|`https://v4.live.maniaplanet.com/ingame/public/titles/download/TMStadium@nadeo.Title.Pack.gbx`||
|TITLE_PACK_FILE|None||
|DEDICATED_CFG|`config.xml`||
|MATCH_SETTINGS|`MatchSettings/matchsettings.xml`||
|SUPERADMIN_NAME|`SuperAdmin`||
|SUPERADMIN_PWD|`SuperAdmin`||
|ADMIN_NAME|`Admin`||
|ADMIN_PWD|`Admin`||
|SERVER_NAME|`My maniaplanet Server`||
|COMMENT|None||
|HIDE_SERVER|`0`|value is 0 (always shown), 1 (always hidden), 2 (hidden from nations)|
|MAX_PLAYERS|`32`||
|SERVER_PASSWORD|None||
|MAX_SPECTATORS|`32`||
|PASSWORD_SPECTATOR|None||
|KEEP_PLAYER_SLOTS|`False`|when a player changes to spectator, should the server keep if player slots/scores etc.. or not|
|LADDER_MODE|`forced`|value between 'inactive', 'forced' (or '0', '1')|
|P2P_UPLOAD|`True`||
|P2P_DOWNLOAD|`False`||
|CALLVOTE_TIMEOUT|`60000`||
|CALLVOTE_RATIO|`0.5`|default ratio. value in [0..1], or -1 to forbid|
|ALLOW_MAP_DOWNLOAD|`True`||
|AUTOSAVE_REPLAYS|`False`||
|AUTOSAVE_VALIDATION_REPLAYS|`False`||
|REFEREE_PASSWORD|None||
|REFEREE_VALIDATION_MODE|`0`|value is 0 (only validate top3 players),  1 (validate all players)|
|USE_CHANGING_VALIDATION_SEED|`False`||
|DISABLE_HORNS|`False`||
|MAX_LATENCY|`0`|0 mean automatic adjustement|
|LADDER_SERVERLIMIT_MIN|None||
|LADDER_SERVERLIMIT_MAX|None||
|CONN_UPLOAD_RATE|`8000`|Kbits per second|
|CONN_DOWNLOAD_RATE|`8000`|Kbits per second|
|ALLOW_SPECTATOR_RELAYS|`False`||
|P2P_CACHE_SIZE|`600`||
|CLIENT_PORT|None||
|BIND_IP_ADDRESS|None||
|USE_NAT_UPNP|None||
|GSP_NAME|None|Game Server Provider|
|GSP_URL|None|If you're a server hoster, you can use this to advertise your services|
|SCRIPTCLOUD_SOURCE|`nadeocloud`|Specify the cloud storage mode for Titles that use it. Can be "localdebug" or "xmlrpc" or "nadeocloud" (default). "nadeocloud" will work only if the creator of the title subscribed to the cloud service|
|BLACKLIST_URL|None||
|GUESTLIST_FILENAME|None||
|BLACKLIST_FILENAME|None||
|MIN_CLIENT_BUILD|None|Only accept updated client to a specific version. ex: 2011-10-06|
|DISABLE_COHERENCE_CHECKS|`False`|disable internal checks to detect issues/cheats, and reject race times|
|DISABLE_REPLAY_RECORDING|`False`|disable replay recording in memory during the game to lower memory usage|
|SAVE_ALL_INDIVIDUAL_RUNS|`False`|Save all the ghosts from the match replay to individual ghost.gbx files, in folder {servername}/Autosaves/Runs_{mapname}/|
|USE_PROXY|`False`||
|PROXY_LOGIN|None||
|PROXY_PWD|None||

## Sources
- https://github.com/PyPlanet/maniaplanet-docker
- http://blog.zot24.com/tips-tricks-with-alpine-docker/
- https://doc.maniaplanet.com/dedicated-server/getting-started
- https://pterodactyl.io/community/config/eggs/creating_a_custom_image.html#creating-the-dockerfile

Lint with :
- https://github.com/hadolint/hadolint
- https://github.com/koalaman/shellcheck


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="2.0.1"></a>
## [2.0.1] - 2022-04-21
### Fixed
- issue [#1](https://git.tools01.noxinmortus.fr/sysadmins/ansible/role-zabbix/issues/1) multiple-download-of-titles


<a name="2.0.0"></a>
## [2.0.0] - 2022-03-16
### Changed
- Major upgrade of... everything


<a name="1.2.3"></a>
## [1.2.3] - 2022-03-16
### Fixed
- WORKDIR issue
- WORKDIR issue


<a name="1.2.2"></a>
## [1.2.2] - 2022-03-16
### Fixed
- Entrypoint incorrect


<a name="1.2.1"></a>
## [1.2.1] - 2022-03-16
### Fixed
- Renamed all Trackmania > ManiaPlanet [skip-ci]


<a name="1.2.0"></a>
## 1.2.0 - 2022-03-16
### Added
- Auto-changelog & auto-tag CI
- docker-pushrm to CI

### Changed
- Update base image to alpine:3.15, update GLIBC to 2.35-r0, add feature to manage user UID & GID
- Update pterodactyl base image to alpine-3.15_glibc-2.34
- Update README.md
- Update alpine version from 3.12 to 3.13

### Fixed
- LICENCE & README


[Unreleased]: https://git.tools01.noxinmortus.fr/sysadmins/ansible/role-zabbix/compare/2.0.1...HEAD
[2.0.1]: https://git.tools01.noxinmortus.fr/sysadmins/ansible/role-zabbix/compare/2.0.0...2.0.1
[2.0.0]: https://git.tools01.noxinmortus.fr/sysadmins/ansible/role-zabbix/compare/1.2.3...2.0.0
[1.2.3]: https://git.tools01.noxinmortus.fr/sysadmins/ansible/role-zabbix/compare/1.2.2...1.2.3
[1.2.2]: https://git.tools01.noxinmortus.fr/sysadmins/ansible/role-zabbix/compare/1.2.1...1.2.2
[1.2.1]: https://git.tools01.noxinmortus.fr/sysadmins/ansible/role-zabbix/compare/1.2.0...1.2.1


entrypoint.sh
#!/bin/sh

# Make sure we use defaults everywhere.
: "${TITLE:=TMStadium@nadeo}"
: "${TITLE_PACK_URL:=https://v4.live.maniaplanet.com/ingame/public/titles/download/TMStadium@nadeo.Title.Pack.gbx}"
: "${DEDICATED_CFG:=config.xml}"
: "${MATCH_SETTINGS:=MatchSettings/matchsettings.xml}"
: "${SERVER_NAME:=My ManiaPlanet Server}"
: "${SERVER_PORT:=2350}"

# We are required to get the public ip if we don't have it in our env currently.
if [ -z ${FORCE_IP_ADDRESS+x} ]; then
   FORCE_IP_ADDRESS=$(wget -4 -qO- http://ifconfig.co)
fi
echo "=> Going to run on forced IP: ${FORCE_IP_ADDRESS} and port: ${SERVER_PORT}"

# Copy the configuration files if not yet copied.
mkdir -pv "${WORKDIR}"/UserData/Config "${WORKDIR}"/UserData/Packs "${WORKDIR}"/UserData/Maps/MatchSettings
## {{{ config.xml
if [ ! -f "${WORKDIR}"/UserData/Config/config.xml ]; then
echo "<?xml version=\"1.0\" encoding=\"UTF-8\"?>
<dedicated>
	<authorization_levels>
		<level>
			<name>${SUPERADMIN_NAME:-SuperAdmin}</name>
			<password>${SUPERADMIN_PWD:-SuperAdmin}</password>
		</level>
		<level>
			<name>${ADMIN_NAME:-Admin}</name>
			<password>${ADMIN_PWD:-Admin}</password>
		</level>
	</authorization_levels>

 	<masterserver_account>
		<login>${MASTERSERVER_ACCOUNT:-}</login>
		<password>${MASTERSERVER_ACCOUNT_PWD:-}</password>
		<validation_key>${MASTERSERVER_KEY:-}</validation_key>
	</masterserver_account>

	<server_options>
		<name>${SERVER_NAME:-}</name>
		<comment>
			${COMMENT:-}
		</comment>
		<hide_server>${HIDE_SERVER:-0}</hide_server>
		<max_players>${MAX_PLAYERS:-32}</max_players>
		<password>${SERVER_PASSWORD:-}</password>
		<max_spectators>${MAX_SPECTATORS:-32}</max_spectators>
		<password_spectator>${PASSWORD_SPECTATOR:-}</password_spectator>
		<keep_player_slots>${KEEP_PLAYER_SLOTS:-False}</keep_player_slots>
		<ladder_mode>${LADDER_MODE:-forced}</ladder_mode>
		<enable_p2p_upload>${P2P_UPLOAD:-True}</enable_p2p_upload>
		<enable_p2p_download>${P2P_DOWNLOAD:-False}</enable_p2p_download>
		<callvote_timeout>${CALLVOTE_TIMEOUT:-60000}</callvote_timeout>
		<callvote_ratio>${CALLVOTE_RATIO:-0.5}</callvote_ratio>
		<callvote_ratios>
			<voteratio command=\"Ban\" ratio="-1"/>
			<!-- commands can be \"Ban\", \"Kick\", \"RestartMap\", \"NextMap\", \"SetModeScriptSettingsAndCommands\"... -->
		</callvote_ratios>
		<allow_map_download>${ALLOW_MAP_DOWNLOAD:-True}</allow_map_download>
		<autosave_replays>${AUTOSAVE_REPLAYS:-False}</autosave_replays>
		<autosave_validation_replays>${AUTOSAVE_VALIDATION_REPLAYS:-False}</autosave_validation_replays>
		<referee_password>${REFEREE_PASSWORD:-}</referee_password>
		<referee_validation_mode>${REFEREE_VALIDATION_MODE:-0}</referee_validation_mode>
		<use_changing_validation_seed>${USE_CHANGING_VALIDATION_SEED:-False}</use_changing_validation_seed>
		<disable_horns>${DISABLE_HORNS:-False}</disable_horns>
		<clientinputs_maxlatency>${MAX_LATENCY:-0}</clientinputs_maxlatency>
		<ladder_serverlimit_min>${LADDER_SERVERLIMIT_MIN:-}</ladder_serverlimit_min>
		<ladder_serverlimit_max>${LADDER_SERVERLIMIT_MAX:-}</ladder_serverlimit_max>
	</server_options>

	<system_config>
		<connection_uploadrate>${CONN_UPLOAD_RATE:-8000}</connection_uploadrate>
		<connection_downloadrate>${CONN_DOWNLOAD_RATE:-8000}</connection_downloadrate>
		<allow_spectator_relays>${ALLOW_SPECTATOR_RELAYS:-False}</allow_spectator_relays>
		<p2p_cache_size>${P2P_CACHE_SIZE:-600}</p2p_cache_size>
		<force_ip_address>${FORCE_IP_ADDRESS:-}</force_ip_address>
		<server_port>${SERVER_PORT:-2350}</server_port>
		<server_p2p_port>${SERVER_P2P_PORT:-3450}</server_p2p_port>
		<client_port>${CLIENT_PORT:-0}</client_port>
		<bind_ip_address>${BIND_IP_ADDRESS:-}</bind_ip_address>
		<use_nat_upnp>${USE_NAT_UPNP:-}</use_nat_upnp>
		<gsp_name>${GSP_NAME:-}</gsp_name>
		<gsp_url>${GSP_URL:-}</gsp_url>
		<xmlrpc_port>${XMLRPC_PORT:-5000}</xmlrpc_port>
		<xmlrpc_allowremote>${XMLRPC_ALLOW_REMOTE:-True}</xmlrpc_allowremote>
		<scriptcloud_source>${SCRIPTCLOUD_SOURCE:-nadeocloud}</scriptcloud_source>
		<blacklist_url>${BLACKLIST_URL:-}</blacklist_url>
		<guestlist_filename>${GUESTLIST_FILENAME:-}</guestlist_filename>
		<blacklist_filename>${BLACKLIST_FILENAME:-}</blacklist_filename>
		<title>${TITLE:-TMStadium@nadeo}</title>
		<minimum_client_build>${MIN_CLIENT_BUILD:-}</minimum_client_build>
		<disable_coherence_checks>${DISABLE_COHERENCE_CHECKS:-False}</disable_coherence_checks>
		<disable_replay_recording>${DISABLE_REPLAY_RECORDING:-False}</disable_replay_recording>
		<save_all_individual_runs>${SAVE_ALL_INDIVIDUAL_RUNS:-False}</save_all_individual_runs>
		<use_proxy>${USE_PROXY:-False}</use_proxy>
		<proxy_login>${PROXY_LOGIN:-}</proxy_login>
		<proxy_password>${PROXY_PWD:-}</proxy_password>
	</system_config>
</dedicated>
" > "${WORKDIR}"/UserData/Config/config.xml
fi
## config.xml }}}
if [ ! -f "${WORKDIR}"/UserData/Maps/MatchSettings/matchsettings.xml ]; then
    cp -v "${TEMPLATE_DIR}"/matchsettings.xml "${WORKDIR}"/UserData/Maps/MatchSettings/matchsettings.xml
fi
if [ ! -f "${WORKDIR}"/UserData/Maps/stadium_map.Map.gbx ]; then
    cp -v "${TEMPLATE_DIR}"/stadium_map.Map.gbx "${WORKDIR}"/UserData/Maps/stadium_map.Map.gbx
fi

# Build TITLE_PACK_FILE variable if it does not exist from TITLE_PACK_URL
if [ -z ${TITLE_PACK_FILE+x} ]; then
  TITLE_PACK_FILE="$(echo "${TITLE_PACK_URL}" | awk -F'/' '{ print $NF }')"
fi

# Download TITLE_PACK_URL if it does not already exist
if [ ! -f "${WORKDIR}"/UserData/Packs/"${TITLE_PACK_FILE}" ]; then
  echo "=> Downloading newest title version to ${TITLE_PACK_FILE}"
  wget ${TITLE_PACK_URL} -nv -O "${WORKDIR}"/UserData/Packs/"${TITLE_PACK_FILE}"
fi

if [ -z ${MASTERSERVER_ACCOUNT+x} ]; then
  echo 'MASTERSERVER_ACCOUNT variable should not be empty if it is not a LAN-only server !'
fi

# Set trap to stop the script proprely when a docker stop is executed
trap : EXIT TERM INT TERM QUIT WINCH

# Start dedicated.
echo "=> At $(pwd) | Starting server, login=${MASTERSERVER_ACCOUNT:-} with additional parameters :"
echo "${@:-None}"

./ManiaPlanetServer "${@}" \
    /nodaemon \
    /forceip="${FORCE_IP_ADDRESS}":"${SERVER_PORT}" \
    /dedicated_cfg=${DEDICATED_CFG} \
    /game_settings=${MATCH_SETTINGS}
